lib/Runtime/Language/SourceTextModuleRecord.cpp

```diff
@@ -1097,42 +1079,189 @@ namespace Js
+                    {
+                        parentModule->IncrementAwaited();
+                    });
+                }
             }
-            END_SAFE_REENTRANT_CALL
         }
         catch (const Js::JavascriptException &err)
         {
-            exception = err.GetAndClear();
+            SetWasEvaluated();
+            SetEvaluating(false);
+            JavascriptExceptionObject *exception = err.GetAndClear();
             Var errorObject = exception->GetThrownObject(scriptContext);
             AssertOrFailFastMsg(errorObject != nullptr, "ModuleEvaluation: null error object thrown from root function");
             this->errorObject = errorObject;
             if (this->promise != nullptr)
             {
                 ResolveOrRejectDynamicImportPromise(false, errorObject, scriptContext, this, false);
-                return scriptContext->GetLibrary()->GetUndefined();
+            }
+            else if (!isRootModule || !scriptContext->GetHostScriptContext()->ReportModuleCompletion(this, errorObject))
+            {
+                JavascriptExceptionOperators::DoThrowCheckClone(exception, scriptContext);
             }
         }
 
-        if (exception != nullptr)
+        return scriptContext->GetLibrary()->GetUndefined();
+    }
+
+    void SourceTextModuleRecord::FinishModuleEvaluation(bool shouldIncrementAwait)
+    {
+        if (WasEvaluated())
         {
-            JavascriptExceptionOperators::DoThrowCheckClone(exception, scriptContext);
+            return;
         }
 
-        if (this->promise != nullptr)
+        CleanupBeforeExecution();
+
+        JavascriptGenerator* gen = static_cast<JavascriptGenerator*> (generator);
+
+        AUTO_NESTED_HANDLED_EXCEPTION_TYPE((ExceptionType)(ExceptionType_OutOfMemory | ExceptionType_JavascriptException));
+        BEGIN_SAFE_REENTRANT_CALL(scriptContext->GetThreadContext())
         {
-            SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(true, this->GetNamespace(), this->GetScriptContext(), this, false);
+            if (gen->IsAsyncModule())
+            {
+                JavascriptPromise* prom = JavascriptAsyncFunction::BeginAsyncFunctionExecution(this->generator);
+                if (this->parentModuleList != nullptr && shouldIncrementAwait)
+                {
+                    parentModuleList->Map([=](uint i, SourceTextModuleRecord* parentModule)
+                    {
+                        parentModule->IncrementAwaited();
+                    });
+                }
+                auto* fulfilled = scriptContext->GetLibrary()->CreateAsyncModuleCallbackFunction(EntryAsyncModuleFulfilled, this);
+                auto* rejected = scriptContext->GetLibrary()->CreateAsyncModuleCallbackFunction(EntryAsyncModuleRejected, this);
+                auto* unused = JavascriptPromise::UnusedPromiseCapability(scriptContext);
+                JavascriptPromise::PerformPromiseThen(prom, unused, fulfilled, rejected, scriptContext);
```

Comment:
Module with top level await execute until the first await then PerformPromiseThen to queue the continuation.